OFEP-single-context-paradigm.md

```diff
@@ -39,12 +39,14 @@ interface Provider {
   //...
 
   // a handler called by the SDK when context is modified
-  onContextChange?(oldContext: EvaluationContext, newContext: EvaluationContext): Promise<void>
+  onContextSet?(oldContext: EvaluationContext, newContext: EvaluationContext): Promise<void>
```

Comment:
Will `oldContext` be an empty object the first time this is called?

Reply:
IMO it's best not to couple initialization to context changes.

And I think we will need to expose a "single context mode", which providers can opt in and implement if they support that.

In this mode, it is best to expose an "initializer" , which performs the first initialization, caching and other provider specifics. 

This way, we can isolate implementation specifics easily.

Reply:
Pseudocode could look like this,

```
# initialization 
OpenFeatureAPI.setSingleContextProvider(MySingleContextProvider, Context)

# Inside setSingleContextProvider implementation
MySingleContextProvider.initialize(context)

# Inside MyProvider.initialize  implementation
// Bulk fetch of feature flags with provided context data
// Cache evaluation for offline/non-IO bound flag providing
```

Reply:
Agreed, @Kavindu-Dodan . I added an initialization function in my [draft spec changes](https://github.com/open-feature/spec/pull/171/files#diff-e688c13af607ec6fef905d28a21cab877987122c01d34af42b874bb98e49b672R159).

Reply:
I've done this: https://github.com/open-feature/ofep/pull/41/commits/4083c83bd37ae4f37dc02155b27b256a6ac937bc

Reply:
Following up from @Kavindu-Dodan suggestion:

Should we _only_ expose a `setProvider` API that requires to pass in both the new provider and a new non-null context? Can we foresee any use-case where users would want to set a new provider without any global context available?

The reason why I am asking: as long as it's possible to set a new provider with global context being null, the `initialize` and `onContextSet` functions would need to pass the current context _optionally_ to the provider (for the languages that support optional types).

The options being (pseudocode):
- Allow `setProvider(provider)` and pass the context optionally in followup functions, e.g. `initialize(context?)`
- Only allow `setProvider(provider, context)` and pass the context strictly not-null in followup functions, e.g. `initialize(context)`

Reply:
> Can we foresee any use-case where users would want to set a new provider without any global context available?

I think this is a valid use case. For example, at the application startup, the provider can be set globally. Yet the context can be empty/null at this stage as context details are unavailable. But once the context is ready/known (ex:- user login, region detection, page visit...), the context can be initialized. 

Contrary to my pseudocode, it makes sense to go with `initialize(context)` followed by `setProvider(provider)`.